trieste/acquisition/function.py

```diff
@@ -512,15 +501,158 @@ def prepare_acquisition_function(
+        mean, cov = model.predict(at)   # [N, L], [N, L]
+        return mean + tf.math.sqrt(cov) * eps[:, None, :]  # [S, N, L]
+
+    return sample
+
+
+# todo the batch size problem
+def batch_latent_sampler_with_reparametrisation(
+    dataset: Dataset,
+    model: ProbabilisticModel,
+    sample_size: int,
+    batch_size: int,
+    observation_shape: ShapeLike,
+    jitter: Union[tf.Tensor, float] = DEFAULTS.JITTER
+) -> Callable[[tf.Tensor], tf.Tensor]:
+    """
+    Returns batch-correlated samples according to the reparametrization trick.
+    The sample size S is determined by eps,
+    the batch size B must be compatible between eps and at,
+    N is the number of batches of query points
+    and L is the number of latent processes.
+    :param dataset:
+    :param model:
+    :param sample_size:
+    :param batch_size:
+    :param observation_shape:
+    :param jitter:
+    :return: a [S, N, B, L] tensor
+    """
+    eps = tf.random.normal(
+        observation_shape + [batch_size, sample_size], dtype=dataset.observations.dtype
+    )  # [L, B, S]
+
+    def sample(at: tf.Tensor) -> tf.Tensor:
+        # todo do we need this shape check?
+        tf.debugging.assert_shapes([(at, [..., sample_size, batch_size, '_'])])
+
+        # todo check this is the same as what victor wrote
+        mean, cov = model.predict(at, full_cov=True)  # [N, B, L], [N, L, B, B]
+        identity = tf.eye(batch_size, dtype=cov.dtype)  # [B, B]
+        cov_cholesky = tf.linalg.cholesky(cov + jitter * identity)  # [N, L, B, B]
+        return mean + tf.transpose(cov_cholesky @ eps, [3, 0, 2, 1])  # [S, N, B, L]
+
+    return sample
+
+
+class BatchMonteCarloExpectedImprovement(SingleModelBuilder[BatchAcquisitionFunction]):
+    """
+    Builder for the Monte_carlo based expected improvement.
+    """
+
+    def __init__(self, num_samples: int, num_query_points: int):
+        super().__init__()
+        self._num_samples = num_samples
+        self._num_query_points = num_query_points
+
+    def build(self, dataset: Dataset, model: ProbabilisticModel) -> BatchAcquisitionFunction:
         """
-        :param datasets: The data from the observer.
-        :param models: The models over each dataset in ``datasets``.
-        :return: A batch acquisition function.
+        :param dataset: The data from the observer.
+        :param model: The model over the specified ``dataset``.
+        :return: The expected improvement function.
         """
+        tf.debugging.assert_shapes([(dataset.observations, (..., 1))])
+
+        mean, _ = model.predict(dataset.query_points)
+        eta = tf.reduce_min(mean, axis=0)
+
+        sampler = batch_latent_sampler_with_reparametrisation(
+            dataset,
+            model,
+            self._num_samples,
+            self._num_query_points,
+            # todo this feels like a hack (that we're pretending to support higher-rank)
+            dataset.observations.shape[-1:]
```

Comment:
This is incorrect I think. We are looking for the number of latent models; we can have a single output but multiple latents. This is the case of the HetGP model (one GP for the mean, another for the variance).

Reply:
it feels to me that we should restrict this `sampler` to `GPflowPredictor`, and update this class to register the number of latent models

Reply:
restricting to `GPflowPredictor` is a significant design change that I want to avoid for a number of reasons

we should chat about this